src/decoder/mp3.rs

```diff
@@ -1,129 +1,917 @@
+///
+/// # Frame-based Processing
+///
+/// MP3 audio is organized into variable-size frames containing compressed audio data.
+/// Each frame is decoded independently, containing 384 (Layer I) or 1152 (Layer II/III)
+/// samples per channel. The decoder maintains the current frame and tracks position
+/// within it for efficient sample-by-sample iteration.
+///
+/// # Seeking Behavior
+///
+/// MP3 seeking accuracy depends on the encoding type:
+/// - **Constant Bitrate (CBR)**: Accurate byte-position-based seeking
+/// - **Variable Bitrate (VBR)**: Approximate seeking with potential drift
+/// - **Average Bitrate (ABR)**: Moderate accuracy depending on variation
+///
+/// # Bitrate Adaptation
+///
+/// The decoder tracks average bitrate over time to improve seeking accuracy,
+/// especially for VBR files where initial estimates may be inaccurate.
+///
+/// # Channel Changes
+///
+/// MP3 frames can theoretically change channel configuration, though this is
+/// rare in practice. The decoder handles such changes dynamically.
+///
+/// # Thread Safety
+///
+/// This decoder is not thread-safe. Create separate instances for concurrent access
+/// or use appropriate synchronization primitives.
+///
+/// # Generic Parameters
+///
+/// * `R` - The underlying data source type, must implement `Read + Seek`
 pub struct Mp3Decoder<R>
 where
     R: Read + Seek,
 {
-    // decoder: SeekDecoder<R>,
-    decoder: Decoder<R>,
-    // what minimp3 calls frames rodio calls spans
-    current_span: Frame,
+    /// The underlying minimp3 decoder, wrapped in Option for seeking operations.
+    ///
+    /// Temporarily set to `None` during stream reset operations for seeking.
+    /// Always `Some` during normal operation and iteration.
+    decoder: Option<Decoder<R>>,
+
+    /// Byte position where audio data begins (after headers/metadata).
+    ///
+    /// Used as the base offset for seeking calculations. Accounts for ID3 tags,
+    /// XING headers, and other metadata that precedes the actual audio frames.
+    start_byte: u64,
+
+    /// Current decoded MP3 frame (what minimp3 calls frames, rodio calls spans).
+    ///
+    /// Contains the raw PCM samples from the current frame. `None` indicates
+    /// either stream exhaustion or that a new frame needs to be decoded.
+    current_span: Option<Frame>,
+
+    /// Current position within the current frame.
+    ///
+    /// Tracks the next sample index to return from the current frame's data.
+    /// When this reaches the frame's sample count, a new frame must be decoded.
     current_span_offset: usize,
+
+    /// Number of audio channels.
+    ///
+    /// Can theoretically change between frames, though this is rare in practice.
+    /// Updated dynamically when frame channel count changes.
+    channels: ChannelCount,
+
+    /// Sample rate in Hz.
+    ///
+    /// Fixed for the entire MP3 stream. Common rates include 44.1kHz (CD quality),
```

Comment:
I dont think we need to know what common rates there are in the comments on a private field.

Reply:
It is usefull to note the sample rate is fixed though.

I would write this doc comment as:
`does not change after decoder creation`